modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,478 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+//
+//                          License Agreement
+//                For Open Source Computer Vision Library
+//
+// Copyright (C) 2020, Huawei Technologies Co., Ltd. All rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Liangqian Kong <chargerKong@126.com>
+//         Longbu Wang <riskiest@gmail.com>
+
+#include <opencv2/core.hpp>
+#include <iostream>
+namespace cv
+{
+    #define CV_QUAT_EPS 1.e-6
+    /**
+     * Quaternion are a number system that extends the complex numbers.
+     * A quaternion is generally represented in the form:
+     * 		\f[\begin{array}{rcl}
+     * 		q &= w + x\boldsymbol{i} + y\boldsymbol{j} + z\boldsymbol{k}\\
+     *		  &= [w, x, y, z]\\
+     *		  &= [w, \boldsymbol{v}]
+     *		  \end{array}.\f]
+     * A unit quaternion is usually represents rotation, which has the form:
+     * 		\f[q = [cos\theta, sin\theta\cdot u_x, sin\theta\cdot u_y, sin\theta\cdot u_z].\f]
+     *
+     * In fact. A general quaternion could also be represented in the form
+     * 		\f[q = ||q||[cos\theta, sin\theta\cdot u_x, sin\theta\cdot u_y, sin\theta\cdot u_z].\f]
+     * where \f$||q||\f$ represents the norm of \f$q\f$
+     * all functions will be degenerate to unit cases when \f$||q|| = 1\f$
+     */
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
+        std::is_same<double, _Tp>::value,
+        _Tp
+        >::type;
+    public:
+        Quat() = default;
+        explicit Quat(const cv::Vec<_Tp, 4> &coeff);
+        Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis, const _Tp qNorm=1);
+        explicit Quat(const cv::Mat &R);
```

Comment:
agreed.

Reply:
Agreed

Reply:
Did you mean Matx？
We think that Matx is not as widely used as mat,  and it can't be transformed from Mat to Matx implicitly, so we retain  it.

Reply:
I think that the true OpenCV style is to use `InputArray` instead of `cv::Mat` as a function argument (if we need such a constructor at all, which is doubtful)

Reply:
I found change cv::Mat to InputArray will raise an ambiguous error, I think this maybe result from the constructor `Quat(const cv::Vec<_Tp, 4> &coeff)` and `Quat(const cv::Vec<_Tp, 3> &rvec)`. So I didn't modify it.

Reply:
I would suggest to remove this ctor.
It can be misused (as mentioned above Vec4 coeff / Vec3 rvec).

If you really want this overload then you may add this:
```
static inline Quat<_Tp> createFromCoeff(InputArray coeff);
static inline Quat<_Tp> createFromRvec(InputArray rvec);

Reply:
At least that'd be good to have a constructor from `cv::Matx<_Tp, 3, 3>`

Reply:
Thanks very much for your code review. @savuor @alalek 

I think it's natural to construct a quaternion by a Vec4. 
Mat to quaternion is a bit more complicated which feels more suitable for implementation in createFrom* function, so we want to remove `explicit Quat(const cv::Mat &R);` and Vec3 rvec,  and add createFromMat and createFromRvec. 

can this be OK?@savuor @alalek